### PR TITLE
Fix caching test failures

### LINKED ISSUE #8045: Windows caching tests failing with FileNotFoundError
```
numba.tests.test_extending.TestCachingOverloadObjmode.test_caching_overload_objmode
numba.tests.test_extending.TestCachingOverloadObjmode.test_check_objmode_cache_ndarray
```

These start to fail at merge of #7965 (e0a73c1). Most likely reason is that the path is too long.

```
numba.tests.test_withlifting.TestLiftObjCaching.test_objmode_caching_call_closure_bad
```
also fails in the buildfarm but cannot be replicated outside of the full run yet.

### PR DESCRIPTION
**Pull request by sklam**
Fix #8045.

### CODE REVIEW

**Comment by sklam** [comment]
BFID: `numba_smoketest_cpu_yaml_84`

**Comment by stuartarchibald** [comment]
> BFID: `numba_smoketest_cpu_yaml_84`

This is passing on all platforms not affected by https://github.com/numba/numba/issues/8043

File: numba/core/caching.py [hunk 1]
```diff
@@ -555,7 +555,8 @@ def _open_for_write(self, filepath):
         Open *filepath* for writing in a race condition-free way (hopefully).
         uuid4 is used to try and avoid name collisions on a shared filesystem.
         """
-        tmpname = '%s.tmp.%d' % (filepath, uuid.uuid4())
+        uid = uuid.uuid4().hex[:16]  # avoid long paths
```
Comment (stuartarchibald) [inline numba/core/caching.py:558]:
```suggestion
        uid = hash(uuid.uuid4())  # avoid long paths
```
makes it shorter?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch and for fixes this. Few minor comments/questions else looks good.

File: numba/core/caching.py (hunk 1 above)
Comment (sklam) [reply numba/core/caching.py:558]: not always:

```python

In [13]: uuid.uuid4().hex[:16]
Out[13]: '2814e0e5e9f24749'

In [14]: hash(uuid.uuid4())
Out[14]: 328160728412730801
```

Comment (stuartarchibald) [reply numba/core/caching.py:558]: I guess I was lucky in my tiny sample size!
```python
In [24]: dd = defaultdict(int)

In [25]: for _ in range(1000000):dd[len(str(hash(uuid.uuid4())))]+=1

In [26]: dd
Out[26]: 
defaultdict(int,
            {19: 565845,
             18: 390817,
             16: 3887,
             17: 39052,
             15: 363,
             14: 32,
             13: 4})
```

File: numba/tests/test_extending.py [hunk 2]
```diff
@@ -1935,8 +1939,10 @@ def test_check_objmode_cache_ndarray(self):
         # Env is missing after cache load.
         cache_dir = temp_directory(self.__class__.__name__)
         with override_config("CACHE_DIR", cache_dir):
-            # Test in local process to populate the cache.
-            self.check_objmode_cache_ndarray()
+            # Run in new process to populate the cache
+            run_in_new_process_in_cache_dir(
+                self.populate_objmode_cache_ndarray_check_cache, cache_dir
+            )
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_extending.py:1945]: Ah I see, thanks.

File: numba/core/caching.py [hunk 3]
```diff
@@ -684,6 +685,8 @@ def _index_key(self, sig, codegen):
         codebytes = self._py_func.__code__.co_code
         if self._py_func.__closure__ is not None:
             cvars = tuple([x.cell_contents for x in self._py_func.__closure__])
+            # Note: cloudpickle serializes a function differently depending
+            #       on how the process is launched; e.g. multiprocessing.Process
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/caching.py:689]: b89199e9323f3f801c35fa0f20524833c18f46c4 commit message explained, thanks.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch.

**Comment by sklam** [comment]
tested as part of #8053 and its all green

**Pull request merged**